Implementation theoretically informed Langevin Dynamics (TILD) KSPACE package into LAMMPS

tabedzki opened the pull request:
**Summary**

Add an additional kspace-based class (TILD) to calculate non-bonded pairwise forces indirectly via a particle-mesh scheme through a gridded density representation. This potential does NOT calculate any Coulombic interactions.

**Related Issue(s)**

None

**Author(s)**

Christian Tabedzki, University of Pennsylvania <cat159@scarletmail.rutgers.edu>
Andrew Santos, Sandia National Lab <a.p.santos11@gmail.com>
Zachariah Vicars, University of Pennsylvania

**Licensing**

By submitting this pull request, we agree, that our contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

As this a new feature, there should be zero backward compatibility problems.

**Implementation Notes**

The technique is created as KSPACE class, is heavily inspired by the PPPM class, (specifically the `pppm.cpp` file), and will need the KSPACE package for compilation. 
The code has been verified against the Riggleman Group's in-house code for this technique. 
Because this implementation is a derivative of the KSPACE class, charges cannot be implemented at the same time as this class. Creating something like `kspace_hybrid` would be a workaround for this. Discussion on this is welcome. 

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system (copies into src; unaware of how to force `kspace` installation as a requirement)
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

Papers that this technique is based on (please note the name of this technique has gone through different iterations and we using the name currently used by the Riggleman Group at the University of Pennsylvania).

Chao, H., Koski, J. & Riggleman, R.  "Solvent vapor annealing in block copolymer nanocomposite films: a dynamic mean field approach" Soft Matter, 13(1) 239-249. (2017) https://doi.org/10.1039/C6SM00770H

Fredrickson, G. H. and Orland, H. Dynamics of polymers: A mean-field theory" The Journal of Chemical Physics 140, 084902 (2014) https://doi.org/10.1063/1.4865911

Grzetic, D. J., Wickman, R. A., and Shi, A.-C., "Statistical dynamics of classical systems: A self-consistent field approach", The Journal of Chemical Physics 140, 244907 (2014)  https://doi.org/10.1063/1.4884825

tabedzki commented:
I do not know the proper way of implementing the `tild` entry for the `src/Makefile`, especially for ensuring that the `KSPACE` package is installed as a prerequisite. Help from the LAMMPS devs would be appreciated. Should this be address on the public forum or in something like the mailing list, the new forum, or the LAMMPS Slack?

tabedzki commented:
One big question for us is whether this line should be included.

https://github.com/tabedzki/lammps/blob/238c541460748dacdff2f6919994261919b38533/src/TILD/tild.cpp#L305

tabedzki reviewed (commented):
For cross_iter, you submitted a change to turn on the potent_type_map. What volume does a particle contribute if we didn't specify a shape for the individual type? Will it have zero volume and zero rho0 contribution while still interacting with other types?

sjplimp commented:
Thanks for submitting this new TILD package to LAMMPS - it's a nice addition.  The source code changes/additions look basically fine.  For the doc pages, however, I want to ask for a reorg that I think will be simpler for users to navigate.
And separate TILD from Coulomb KSpace conceptually.

a) Create a new kspace_tild.rst file and reference it from the kspace.rst and kspace_modify.rst doc pages.  Similar
to how dump_image.rst is referenced from the dump.rst doc page, as a separate command, b/c it's syntax is so different.
Put all the info about both the TILD-related kspace_style and kspace_modify commands on this page.  Also, this becomes a stand-alone page where you can explain the TILD model, its use cases and guidelines (e.g. use Langevin too), and citations to the user so they understand why/when it is good to use.  As well as its syntax.  Add a pointer to the new examples/tild dir as well.

b) On the kspace_modify.rst doc page, just have one "tild" keyword with a syntax similar to the scafacos keyword where its internal options are the 1st arg.  So tild/set_rho0 2.0 becomes tild set_rho0 2.0 in all cases (this implies a similar code change as well in tild.cpp).  But don't document any of the TILD options on the kspace_modify.rst page.  Rather point the user to the kspace_tild.rst doc page for the details.

c) Similarly, put all the TILD info on the kspace_style.rst doc page on the new kspace_tild.rst doc page instead.  Just like the dump doc page has no info about the dump image variant, except a link to it's own doc page.

Separate small issues:

src/TILD/README: add your 2 names, not just 2 email addresses
for the kspace_tild doc page, change the grid_resolution param to gridsize and say that is has (distance units)
change KSpace::pair_check() to just add a first line: if (tildflag) return;
for your Q about how to insure the KSPACE package is also installed if TILD is installed.
   I defer to @akohlmey Axel - the issue is not a dervied child class, but that tild.cpp needs the FFT files from KSpace

sjplimp commented:
Also 2 more points:

(1) Not understanding TILD, the doc page info on the 3 itype itype and itype jtype options is still not clear.  E.g. which need to be specified, which have defaults.  The bottom of the page also mentions tild/mix but it's not otherwise mentioned?  Possibly some of these 3 variants overlay or override the others?  This seems very pair_coeff like, but you previously convinced me it was not quite the same.

(2) Eventually I think the tild ave/grid and write_grid options should be removed from tild.cpp, in favor of new fix ave/grid
and/or dump grid commands.  But they are fine to leave there until that happens.

akohlmey commented:
> for your Q about how to insure the KSPACE package is also installed if TILD is installed.
> I defer to @akohlmey Axel - the issue is not a dervied child class, but that tild.cpp needs the FFT files from KSpace

The PHONON package also requires the KSPACE package to be installed to have access to the FFT wrapper. This is done in `src/PHONON/Install.sh` (look for the check on fft3d_wrap.h) and `cmake/CMakeLists.txt` (look for use of `pkg_depends()` function)

tabedzki commented:
Hi all. I am working on the aspect that deals with dependencies for traditional make. I've copied and modified `src/PHONON/Install.sh` to `src/TILD/Install.sh` (see below) as well as added the TILD package to the list of packages in the `Makefile` and to `Depend.sh`. 

I have Uninstall working when I uninstall the `KSPACE` package, but not the functionality of reinstalling it when the `yes-kspace` is activated after a `yes-tild`. PHONON automatically reinstalls after a `yes-kspace; yes-phonon;no-kspace; yes-kspace`

Any idea where I am missing a step about including this extra dependency for the re-install process?


```zsh
(base) ➜  src git:(tild_refactor) ✗ make yes-phonon; make yes-kspace; make yes-phonon; make no-kspace; make yes-kspace; make no-phonon; make no-kspace
Installing package phonon
Must install KSPACE package with PHONON
Installing package kspace
Installing package phonon
Uninstalling package kspace
  updating package PHONON
Installing package kspace
  updating package PHONON
  updating src/fix_phonon.cpp
  updating src/fix_phonon.h
Uninstalling package phonon
Uninstalling package kspace
(base) ➜  src git:(tild_refactor) ✗ make yes-tild; make yes-kspace; make yes-tild; make no-kspace; make yes-kspace; make no-tild; make no-kspace
Installing package tild
Must install KSPACE package with TILD
Installing package kspace
Installing package tild
Uninstalling package kspace
  updating package TILD
Installing package kspace
Uninstalling package tild
Uninstalling package kspace
```


`src/TILD/Install.sh`
```sh
# Install/unInstall package files in LAMMPS
# mode = 0/1/2 for uninstall/install/update

mode=$1

# enforce using portable C locale
LC_ALL=C
export LC_ALL

# arg1 = file, arg2 = file it depends on

action () {
  if (test $mode = 0) then
    rm -f ../$1
  elif (! cmp -s $1 ../$1) then
    if (test -z "$2" || test -e ../$2) then
      cp $1 ..
      if (test $mode = 2) then
        echo "  updating src/$1"
      fi  
    fi  
  elif (test -n "$2") then
    if (test ! -e ../$2) then
      rm -f ../$1
    fi  
  fi  
}

# TILD uses the parallel FFT wrapper used in PPPM,
# so we must require the KSPACE package to be installed.

if (test $1 = 1) then
  if (test ! -e ../fft3d_wrap.h) then
    echo "Must install KSPACE package with TILD"
    exit 1
  fi  
fi

# list of files with optional dependcies

action tild.cpp fft3d_wrap.h
action tild.h fft3d_wrap.h

```


Modified portion of `src/Depend.sh`
```sh
if (test $1 = "KSPACE") then
  depend CG-SDK
  depend CORESHELL
  depend DIELECTRIC
  depend GPU 
  depend KOKKOS
  depend OPT 
  depend OPENMP
  depend INTEL
  depend PHONON
  depend FEP 
  depend TILD
fi
```



Snippet of `src/Makefile`

```
PACKAGE = \
        adios \
        asphere \
        awpmd \
        bocs \
        body \
---
        kim \
        kokkos \
        kspace \
        latboltz \
        latte \
        machdyn \
---
        atc \
        dielectric \
        ml-iap \
        phonon \
        tild
```

akohlmey commented:
> Any idea where I am missing a step about including this extra dependency for the re-install process?

There are two steps that handle dependencies between packages for the GNU Make build system. Install.sh is one, the other is the other is the `src/Depend.sh` script. Please note that dependencies also need to be handled correspondingly for the CMake build procedure.

akohlmey commented:
Please also make certain that your PR branch is up-to-date with upstream.

tabedzki commented:
Thanks Axel for those points. The cmake has been verified; I was checking the GNU make dependency. I believe I've addressed both the `Install.sh` and the `src/Depend.sh` scripts as posted above but the unexpected behavior still persists.

akohlmey commented:
> Thanks Axel for those points. The cmake has been verified; I was checking the GNU make dependency. I believe I've addressed both the `Install.sh` and the `src/Depend.sh` scripts as posted above but the unexpected behavior still persists.

When I check out the PR branch I see no changes to `src/Depend.sh` and `src/version.h` says "31 August 2021".

tabedzki commented:
I've made those changes on my local machine; likewise I haven't updated to the newest version of LAMMPS since I wanted to address some issues. I'll revisit this once the PR changes from draft to ready for review.

tabedzki commented:
I've updated the branch to the newest version of LAMMPS and have included the `src/TILD/Install.sh` and updated the `src/Depend.sh`, but I still get slightly different behavior than the example behavior from PHONON. The TILD package gets uninstalled PHONON when KSPACE is uninstalled but it does not get re-installed as one would expect. 
@akohlmey  whenever you get the chance, can you spot the mistake I'm making?

```
(base) ➜  src git:(tild_refactor) make yes-kspace
Installing package kspace
(base) ➜  src git:(tild_refactor) make yes-phonon
Installing package phonon
(base) ➜  src git:(tild_refactor) make yes-tild
Installing package tild
(base) ➜  src git:(tild_refactor) make no-kspace
Uninstalling package kspace
  updating package PHONON
  updating package TILD
(base) ➜  src git:(tild_refactor) make yes-kspace
Installing package kspace
  updating package PHONON
  updating src/fix_phonon.cpp
  updating src/fix_phonon.h
(base) ➜  src git:(tild_refactor) ls tild.cpp fix_phonon.cpp
ls: cannot access 'tild.cpp': No such file or directory
fix_phonon.cpp
```

akohlmey commented:
> @akohlmey whenever you get the chance, can you spot the mistake I'm making?

The mistake is in the interpretation of the output. The PHONON and TILD packages get **updated**. While for TILD that will remove the only style that is included, for PHONON it will *only* remove fix phonon, but leave the dynamical_matrix and third_order commands in place. You can verify this by typing `make pi` which will show the PHONON package as still installed (since it was not completely removed). For TILD this is obviously not possible. Hence the behavior is (mostly) correct. The only issue is that the PHONON package should no longer refuse to be installed without KSPACE, it should just skip installing fix phonon.

akohlmey commented:
The inconsistent behavior of the PHONON package is corrected in commit 88f8e41702c56e2a3a55cf162260890669ed17d2 in PR #3064

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -2699,3 +2823,29 @@ void TILD::ave_grid()
+    if (intrxn.type != NONE ){
+      density_flags[intrxn.i+1] = density_flags[intrxn.j+1] = 1;
+      potent_type_map[0][intrxn.i+1][intrxn.j+1] = potent_type_map[0][intrxn.j+1][intrxn.i+1] = 0;
+      potent_type_map[nstyles][intrxn.i+1][intrxn.j+1] = potent_type_map[nstyles][intrxn.j+1][intrxn.i+1] = 1;
```
See other comment. The last style (which has an index of nstyles) is the manually specific cross-interaction

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -247,10 +253,8 @@ void TILD::init()
-
-  
+  triclinic = domain->triclinic;
+  pair_check();
```
That's an idea for now until kspace gets updated to allows for a charge and TILD simultaneously.

tabedzki commented on src/TILD/tild.h:
```diff
@@ -64,6 +64,7 @@ class TILD : public KSpace{
   // For future kspace_hybrid
   int nstyles;                  // # of sub-styles For future kspace_hybrid
   int **setflag;                 // 0/1 = whether each i,j has been set
+  int *density_flags; 
```
This isn't near referencing the line with `#include <bits/stdc++.h>`. That being said, I removed this line and things still worked. Will remove the line in a new commit.

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -2423,57 +2467,137 @@ void TILD::ev_calculation(const int loc, const int itype, const int jtype) {
+    utils::logmesg(lmp, mesg);
+    mesg.clear();
+
+    if (set_rho0 > 0) {
```
Good idea about including more TILD in the error messages.

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -2423,57 +2467,137 @@ void TILD::ev_calculation(const int loc, const int itype, const int jtype) {
+      } else if (total_sphere_vol > 0 && set_rho0 > 0) {
+        if (modified_rho0 != 0.f) {
+          std::string mesg_mod = fmt::format(
+              "Modified NP rho to {:.6f} to achieve user specified rho.\n", modified_rho0);
```
I have modified the language to be more clear. Will come in a separate update.
Now says
```
 std::string mesg_mod = fmt::format(
              "Modified NP rho set to {:.6f} for all particles to achieve user specified rho.\n", modified_rho0);
```

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -2423,57 +2467,137 @@ void TILD::ev_calculation(const int loc, const int itype, const int jtype) {
+      }
+    }
+
+  if (set_rho0 <= 0.0f){
```
done

tabedzki commented on src/TILD/tild.cpp:
```diff
@@ -2423,57 +2467,137 @@ void TILD::ev_calculation(const int loc, const int itype, const int jtype) {
+      }
+    }
+  } else {
+    MPI_Allreduce(&total_sphere_vol, &temp_double, 1, MPI_DOUBLE, MPI_SUM, world);
```
done

tabedzki commented:
@akohlmey  would you kindly help with the documentation debugging? Following Steve's suggestion, I want to have a separate `kspace_modify.rst`. I don't know if the index for `tild` should then be included in the separate file or in `kspace_style.rst`?

I also have the following warnings even though I _thought_ I was following the intended procedure for each of the respective warnings.

```
Package TILD missing in Packages_details.rst
Found 1 issue(s) with package lists
```

```
Command style entry dynamical_matrix is missing or incomplete in Commands_all.rst
Command style entry third_order is missing or incomplete in Commands_all.rst
KSpace style entry tild is missing or incomplete in Commands_kspace.rst
Found 3 issue(s) with style lists
kspace_style index entry tild is missing
Found 1 issue(s) with style index
```

(The `dynamical_matrix` and `third_order` has been since corrected in LAMMPS; I've pulled in `lammps/develop` locally and this doesn't affect my issue so I've not pushed the additional merge.)

akohlmey commented:
> @akohlmey would you kindly help with the documentation debugging?

The easiest would be you ignore those warnings. Push when you are done and happy and then ping me again. I'll then pull the breanch and fix up what needs fixing (there almost always is something, even when people do an excellent job).

tabedzki commented:
Sounds good. Thanks

tabedzki commented:
I don't know if I missed it in another `.rst` file, but I didn't see a way of how to link to the `examples/tild` folder. Any help on that would be much appreciated.

apsantos commented on doc/src/kspace_tild.rst:
```diff
@@ -234,41 +220,41 @@ The current interaction styles used in *tild cross-interaction* are
 1. *gaussian* : :math:`\sigma` (distance units)
-2. *gaussian-erfc* : :math:`\sigma`\ , :math:`R_p`, :math:`\xi` (all in distance units)
+2. *gaussian_erfc* : :math:`\sigma`\ , :math:`R_p`, :math:`\xi` (all in distance units)
 3. *erfc* : :math:`R_p`\ , :math:`\xi` (both in distance units)
```
As you say in the note, we ask them for sigma squared, not sigma.  That should probably be reflected in this part too, for example

```suggestion
1. *gaussian* : :math:`\sigma^2` (distance:math:`^2` units)
 2. *gaussian_erfc* : :math:`\sigma^2`\ , :math:`R_p`, :math:`\xi` (distance units except for :math:`\sigma^2` )
 3. *erfc* : :math:`R_p`\ , :math:`\xi` (both in distance units)
```

tabedzki replied:
In the code, it takes in the sigma and squares the value. See line 1560. I'm open to a discussion about whether that should change.

apsantos replied:
I thought line 1560 is called with `tild shape`. For `tild cross-interaction` `calc_cross_work` assumes it is squared, right?

apsantos commented on doc/src/kspace_tild.rst:
```diff
@@ -175,32 +161,32 @@ The current shape function styles used in *tild shape* are
 
 The *tild set_rho0* keyword is used to set the TILD density which is calculated
 separately from any other density in LAMMPS. Each defined `gaussian` shape
 particle has a mass of 1, each defined `erfc` shape has a density of 
```
I think mass can be anything.  rho0 is really of a number density.
```suggestion
The *tild set_rho0* keyword is used to set the TILD number density which is calculated
 separately from any other density in LAMMPS. Each defined `gaussian` shape
 particle contributes 1, while each defined `erfc` shape has a density of 
```

apsantos commented on src/TILD/tild.cpp:
```diff
@@ -0,0 +1,2853 @@
+      density_flags[intrxn.i + 1] = density_flags[intrxn.j + 1] = 1;
+      potent_type_map[0][intrxn.i + 1][intrxn.j + 1] =
+          potent_type_map[0][intrxn.j + 1][intrxn.i + 1] = 0;
+      potent_type_map[nstyles][intrxn.i + 1][intrxn.j + 1] =
```
should this ben styles or, should it be `intrxn.type` or 
```istyle = (intrxn.i * (2 * ntypes - intrxn.i - 1)) / 2 + intrxn.j;
potent_type_map[istyle][intrxn.i + 1][intrxn.j + 1] =
potent_type_map[istyle][intrxn.j + 1][intrxn.i + 1] = 1;
```
?

tabedzki replied:
the nstyles is the specified version. but I think we should have a command that overwrites the specifications between two styles directly, which is something I hadn't considered previously

akohlmey commented:
@tabedzki @apsantos this has been sitting idle now for 9 months and would need some significant updates to be compatible with recent changes to the core code (e.g. the grid refactor, updated error message conventions etc.).
I am closing this now. Please resubmit when you have updated and are closer to be ready for merging.

apsantos commented:
Thanks for letting us know Axel.

Andrew

On Sun, Jan 8, 2023 at 8:58 PM Axel Kohlmeyer ***@***.***>
wrote:

> Closed #2947 <https://github.com/lammps/lammps/pull/2947>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/2947#event-8178212849>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4J2FU5VEOLGZFP2C7KX6LWROLHBANCNFSM5EFP3IKQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

akohlmey commented:
For some information about the grid refactoring and necessary changes you can have a look at: https://docs.lammps.org/Developer_updating.html#refactored-grid-communication-using-grid3d-grid2d-classes-instead-of-gridcomm and https://docs.lammps.org/Developer_grid.html
